### PR TITLE
Update to react-native@0.49 and flow-bin@0.53

### PR DESCRIPTION
**Pull request by Ashoat**
This PR was way harder than I initially thought it would be. Took about a week of full-time work and is 47 commits large.

The goal here was to upgrade us to RN 0.49, which requires upgrading to Flow 0.53. Flow 0.53 completely revamps React types, which exposed a whole ton of (almost entirely) legitimate type errors. I've fixed almost all of these errors.

I ended up giving up on [two errors](https://github.com/react-community/react-navigation/commit/d5c410ceff1d4d30521293d9d0632da435fcc08b), which I am pretty sure are Flow bugs, but don't have the time to debug/report at this point.

I also needed to silence the errors coming from `react-native-gesture-handler` in NavigationPlayground, which is required by Expo.

### CODE REVIEW

**Comment by kelset** [comment]
Note to self for Release Notes: this PR also updates: "react-native-tab-view": "^0.0.70"

**Comment by Ashoat** [comment]
Heads-up, I'm seeing some issues integrating this into my RN project. Don't merge yet!

**Comment by davepack** [comment]
@Ashoat what are you seeing? I rebased my pr onto this and it seems to be working fine, both in navigation playground and a separate project.

**Comment by Ashoat** [comment]
Most of the errors I was seeing are fixed by the last two commits, but I'm still seeing one in my project:

```
Error: node_modules/react-navigation/src/createNavigationContainer.js:41
 41:   S: NavigationState,
          ^^^^^^^^^^^^^^^ object type. This type is incompatible with
                    v----------------------------------
 40: export default function createNavigationContainer<
 41:   S: NavigationState,
 42:   A: *,
...:
212: }
     ^ some incompatible instantiation of `S`
```

**Comment by Ashoat** [comment]
Okay, everything is typechecking on my side now!

File: examples/NavigationPlayground/flow-typed/npm/expo_vx.x.x.js [hunk 1]
```diff
@@ -254,6 +314,14 @@ declare module 'expo/src/sensor/__tests__/Accelerometer-test' {
   declare module.exports: any;
 }
 
+declare module 'expo/src/sensor/__tests__/DeviceMotion-test' {
```
Comment (matthargett) [inline examples/NavigationPlayground/flow-typed/npm/expo_vx.x.x.js]: why would expo's tests need to be flow-typed?

Comment (Ashoat) [reply examples/NavigationPlayground/flow-typed/npm/expo_vx.x.x.js]: `flow-typed` creates these autogenerated stubs for any library it doesn't have a libdef for. I'm not entirely sure why, but the comment in the file seems to indicate it's to help people start the "gradual typing" process

File: examples/ReduxExample/package.json [hunk 2]
```diff
@@ -22,17 +22,17 @@
     ]
   },
   "dependencies": {
-    "expo": "^21.0.1",
+    "expo": "^22.0.0",
     "prop-types": "^15.5.10",
-    "react": "16.0.0-alpha.12",
-    "react-native": "~0.48.4",
+    "react": "16.0.0-beta.5",
+    "react-native": "https://github.com/expo/react-native/archive/sdk-22.0.1.tar.gz",
```
Comment (davepack) [inline examples/ReduxExample/package.json]: I don't think we should depend on the Expo fork here, just vanilla react-native

Comment (Ashoat) [reply examples/ReduxExample/package.json]: Yeah sorry, I ended up changing it to `^0.49.3` in a later commit

Comment (spencercarli) [reply examples/ReduxExample/package.json]: Looks like it was changed in the `NavigationPlayground` example but not this one.

File: package.json [hunk 3]
```diff
@@ -53,7 +53,7 @@
     "path-to-regexp": "^1.7.0",
     "prop-types": "^15.5.10",
     "react-native-drawer-layout-polyfill": "^1.3.2",
-    "react-native-tab-view": "^0.0.69"
+    "react-native-tab-view": "^0.0.70"
```
Comment (spencercarli) [inline package.json]: Has this been checked to make sure the breaking changes in this release didn't break any of the tab logic in the app?

Comment (spencercarli) [reply package.json]: Just checked the code and looks like we're in the clear 👍

Comment (kelset) [reply package.json]: To give you some context, it was a [necessary update](https://github.com/react-community/react-navigation/issues/2771#issuecomment-339034027) ;) And yeah it works 👌

File: examples/ReduxExample/package.json (hunk 2 above)
Comment (Ashoat) [reply examples/ReduxExample/package.json]: You're totally right, my bad

**Comment by spencercarli** [comment]
🎉

**Comment by kesha-antonov** [comment]
Nice job! Thank you!

**Pull request merged**